HunYuan opensource

yjc9696 opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes # (issue)
This PR primarily aims to add support for the Hunyuan series of models in inference. We noticed that previous Hunyuan models relied on trust_remote_code for inference, which makes version maintenance difficult and often leads to outdated inference code. To address this, we are integrating the inference code into the Transformers library to support continuous updates for future open-source releases.

The submitted code includes the inference implementations for both hunyuan_v1_dense and hunyuan_v1_moe, along with their corresponding configurations and tokenizers.

For unit testing, we added a single-sample test for the hunyuan_v1_moe model using tencent/Hunyuan-A13B-Instruct. Unfortunately, the hunyuan_v1_dense model is not yet officially open-sourced, so we currently lack a testable model for it，we will update upon model release.
This is my first PR submission. After carefully studying the Contribute to 🤗 Transformers guide, I've modified my code to pass all make fixup checks.

I'd greatly appreciate any feedback if additional changes or improvements are needed - please don't hesitate to point them out!

## Before submitting
- [Y] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [Y] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker
- vision models: @amyeroberts, @qubvel
- speech models: @eustlb
- graph models: @clefourrier

Library:

- flax: @gante and @Rocketknight1
- generate: @zucchini-nlp (visual-language models) or @gante (all others)
- pipelines: @Rocketknight1
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @zach-huggingface, @SunMarc and @qgallouedec
- chat templates: @Rocketknight1

Integrations:

- deepspeed: HF Trainer/Accelerate: @SunMarc @zach-huggingface
- ray/raytune: @richardliaw, @amogkam
- Big Model Inference: @SunMarc
- quantization (bitsandbytes, autogpt): @SunMarc @MekkCyber

Documentation: @stevhliu

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @Rocketknight1
- PyTorch: See Models above and tag the person corresponding to the modality of the example.
- TensorFlow: @Rocketknight1

 -->

Commit:
merge opensource_hunyuan

mingjihantencent committed:
add head_dim

mingjihantencent committed:
fix assertion error

Commit:
fix seen_tokens

Commit:
ready_for_upstream (merge request !17)

Squash merge branch 'ready_for_upstream' into 'main'

* fix configuration type&docstring
* fix style

Commit:
ready_for_upstream (merge request !18)

Squash merge branch 'ready_for_upstream' into 'main'
* add doc
* fix testcode
* fix configuration type&docstring

yjc9696 committed:
rename base model

yjc9696 committed:
remove assert

yjc9696 committed:
Merge branch 'main' into hunyuan_opensource

yjc9696 committed:
Merge branch 'main' into hunyuan_opensource

mingjihantencent committed:
update

Commit:
remove tiktoken

yjc9696 committed:
Merge pull request #1 from yjc9696/hunyuan_mingji_fix

fix usable_length API

mingjihantencent committed:
update

yjc9696 committed:
Merge pull request #2 from yjc9696/hunyuan_opensource_mingji_fix_args

update

yjc9696 committed:
Merge branch 'main' into hunyuan_opensource

mingjihantencent committed:
fix moe and code style (#3)

* update

* fix format

* update

* revert makefile

yjc9696 committed:
fix moe config

ArthurZucker reviewed (commented):
Hey! the main point of guidance is to properly isolate the differences in architecture with your model and other models supported in the library!  This way you can use inheritance to write the modeling code with: https://huggingface.co/docs/transformers/en/modular_transformers ! 
🤗

yjc9696 committed:
fix numel()

yjc9696 committed:
remove prepare_inputs_for_generation

yjc9696 committed:
fix kv_seq_len

yjc9696 committed:
Merge branch 'main' into hunyuan_opensource

yjc9696 committed:
Merge branch 'main' into hunyuan_opensource

yjc9696 committed:
add docs/toctree

ArthurZucker reviewed (commented):
Thanks for the PR! 
Before I can review the modeling code, could you isolate the differences between your model and most recent models using this tutorial to write the modeling code?! 
 https://huggingface.co/docs/transformers/en/modular_transformers.

There are many examples across the codebase on how to implement this for example `DeepSeek`'s `modular_deepseek_v3` and etc!

yjc9696 committed:
remove unused paramter&add licence

* add licence

* remove unused paramter

* fix code

yjc9696 commented on src/transformers/models/hunyuan_v1_dense/configuration_hunyuan_v1_dense.py:
```diff
@@ -0,0 +1,231 @@
+        use_cla (`bool`, *optional*, defaults to `False`):
+            Whether to use CLA in attention
+        cla_share_factor (`int`, *optional*, defaults to 1):
+            The share factor of CLA
```
we remove some unused config

yjc9696 commented on src/transformers/models/hunyuan_v1_dense/configuration_hunyuan_v1_dense.py:
```diff
@@ -0,0 +1,231 @@
+# coding=utf-8
```
fix done

ArthurZucker commented on src/transformers/models/hunyuan_v1_dense/modeling_hunyuan_v1_dense.py:
```diff
@@ -0,0 +1,1551 @@
+        return self.weight * hidden_states.to(input_dtype)
+
+
+ALL_LAYERNORM_LAYERS.append(HunYuanRMSNorm)
```
```suggestion
```

ArthurZucker commented on src/transformers/models/hunyuan_v1_dense/modeling_hunyuan_v1_dense.py:
```diff
@@ -0,0 +1,1551 @@
+        self.register_buffer("sin_cached", (emb.sin() * _mscale).to(dtype), persistent=False)
+
+
+def rotate_half(x):
+    """Rotates half the hidden dims of the input."""
+    x1 = x[..., : x.shape[-1] // 2]
+    x2 = x[..., x.shape[-1] // 2 :]
+    return torch.cat((-x2, x1), dim=-1)
+
+
+def apply_rotary_pos_emb(q, k, cos, sin, position_ids, unsqueeze_dim=1):
+    """Applies Rotary Position Embedding to the query and key tensors.
+
+    Args:
+        q (`torch.Tensor`): The query tensor.
+        k (`torch.Tensor`): The key tensor.
+        cos (`torch.Tensor`): The cosine part of the rotary embedding.
+        sin (`torch.Tensor`): The sine part of the rotary embedding.
+        position_ids (`torch.Tensor`):
+            The position indices of the tokens corresponding to the query and key tensors. For example, this can be
+            used to pass offsetted position ids when working with a KV-cache.
+        unsqueeze_dim (`int`, *optional*, defaults to 1):
+            The 'unsqueeze_dim' argument specifies the dimension along which to unsqueeze cos[position_ids] and
+            sin[position_ids] so that they can be properly broadcasted to the dimensions of q and k. For example, note
+            that cos[position_ids] and sin[position_ids] have the shape [batch_size, seq_len, head_dim]. Then, if q and
+            k have the shape [batch_size, heads, seq_len, head_dim], then setting unsqueeze_dim=1 makes
+            cos[position_ids] and sin[position_ids] broadcastable to the shapes of q and k. Similarly, if q and k have
+            the shape [batch_size, seq_len, heads, head_dim], then set unsqueeze_dim=2.
+    Returns:
+        `tuple(torch.Tensor)` comprising of the query and key tensors rotated using the Rotary Position Embedding.
+    """
+    cos = cos[position_ids].unsqueeze(unsqueeze_dim)
+    sin = sin[position_ids].unsqueeze(unsqueeze_dim)
+
+    q_embed = (q * cos) + (rotate_half(q) * sin)
+    k_embed = (k * cos) + (rotate_half(k) * sin)
+    return q_embed, k_embed
```
this is litterly the same as well

ArthurZucker commented on src/transformers/models/hunyuan_v1_dense/modeling_hunyuan_v1_dense.py:
```diff
@@ -0,0 +1,1551 @@
+    return q_embed, k_embed
+
+
+class HunYuanMLP(nn.Module):
+    """
+    使用 SwiGLU 的 MLP
+    """
+
+    def __init__(self, config: HunYuanDenseV1Config, layer_idx=None, is_shared_mlp=False):
+        super().__init__()
+        self.config = config
+        self.layer_idx = layer_idx
+        self.hidden_size = config.hidden_size
+        self.hidden_act = config.hidden_act
+        self.intermediate_size = config.intermediate_size
+        if self.hidden_act != "silu":
+            warnings.warn("We support SwiGLU only.", category=UserWarning)
+
+        self.gate_proj = nn.Linear(self.hidden_size, self.intermediate_size, bias=False)
+        self.up_proj = nn.Linear(self.hidden_size, self.intermediate_size, bias=False)
+        self.down_proj = nn.Linear(self.intermediate_size, self.hidden_size, bias=False)
+        self.act_fn = ACT2FN[config.hidden_act]
+
+    def forward(self, x):
+        down_proj = self.down_proj(self.act_fn(self.gate_proj(x)) * self.up_proj(x))
+        return down_proj
```
this as well, we don't need the warning

ArthurZucker commented on src/transformers/models/hunyuan_v1_dense/modeling_hunyuan_v1_dense.py:
```diff
@@ -0,0 +1,1551 @@
+
+        bsz, q_len, _ = hidden_states.size()
+
+        if self.attention_type == "cross" and kv_states is not None and isinstance(kv_states, tuple):
```
see my commet about self and cross

ArthurZucker commented on src/transformers/models/hunyuan_v1_dense/modeling_hunyuan_v1_dense.py:
```diff
@@ -0,0 +1,1551 @@
+        return attn_output, attn_weights, past_key_value, (orig_key_states, orig_value_states)
+
+
+class HunYuanFlashAttention2(HunYuanAttention):
```
we no longer add flash and sdpa support with this! We use the `ALL_ATTENTION_FUNCTIONS` which allow for transformers backend in vlllm

ArthurZucker reviewed (commented):
Hey! I cannot stress this enough: please isolate the differences it would help a lot! 
Same for the MoE we have many moe in transformers, the implementations are all standardize to help isolate the differences. 
Even if there are no diff, I am happy to still add the `HunyuanForCausalLM` and etc, as long as we know what it is similar to!

yjc9696 committed:
dense modular

update import

fix

fix

use mistralmodel

fix qknorm

add sliding_window

make style

fix

dense done

hunyuan moe

fix import

fix modular

fixup

fixup

yjc9696 committed:
update model path

yjc9696 committed:
fix mlp_bias

yjc9696 committed:
merge main

yjc9696 committed:
fix modular

yjc9696 committed:
Merge branch 'main' into hunyuan_opensource

yjc9696 committed:
Fix modeling (#5)

* fix attention

* use llamamodel

* fix code

ArthurZucker commented:
Feel free to ping me when you need another review!

yjc9696 commented:
> Feel free to ping me when you need another review!

I noticed that the recent updates to the main branch involving Llama caused some errors in my code, but I've resolved them all now. Could you please re-review the changes?
However, I'm not sure why a few tests are failing, as I don't see any error messages displayed.

yjc9696 committed:
Fix qk (#6)

* fix qk_norm

* fix

* fix modual

yjc9696 commented:
@ArthurZucker I need some help with the tests. They keep showing failures, but when I check the details, there are no errors—only warnings. Could you please help me figure out how to resolve this?Also, I’ve made some minor adjustments to the code for qk_norm. Would you mind reviewing it again ?

ArthurZucker commented on src/transformers/models/hunyuan_v1_moe/modular_hunyuan_v1_moe.py:
```diff
@@ -0,0 +1,358 @@
+    expert_mask_aux = expert_mask.max(dim=-2)[0]
+    tokens_per_group_and_expert = torch.mean(expert_mask_aux.float(), dim=-2)
+    router_prob_per_group_and_expert = torch.mean(gates.float(), dim=-2)
+    l_aux = num_experts**2 * torch.mean(tokens_per_group_and_expert * router_prob_per_group_and_expert)
```
we don't compute the auxilary loss here usually!

ArthurZucker commented on src/transformers/models/hunyuan_v1_moe/modular_hunyuan_v1_moe.py:
```diff
@@ -0,0 +1,358 @@
+    return [l_aux, exp_capacity_rate], combine_weights, dispatch_mask, exp_counts
+
+
+def top1gating(logits: Tensor, random_routing_dropped_token: bool = False):
```
we need to keep only the gating that is actually used!

yjc9696 replied:
only keep took,remove top1

ArthurZucker commented on src/transformers/models/hunyuan_v1_moe/modular_hunyuan_v1_moe.py:
```diff
@@ -0,0 +1,358 @@
+    new_mask1 = mask1 * torch.zeros_like(mask1).scatter_(0, top_idx, 1)
+    mask1 = new_mask1
+    mask1_bk = mask1
+    if random_routing_dropped_token:
```
was this used in the released model?

yjc9696 replied:
already remove

ArthurZucker commented on src/transformers/models/hunyuan_v1_moe/modular_hunyuan_v1_moe.py:
```diff
@@ -0,0 +1,358 @@
+        self.layer_idx = layer_idx
+        self.moe_topk = config.moe_topk
+        self.num_experts = config.num_experts if isinstance(config.num_experts, int) else config.num_experts[layer_idx]
+        if config.use_mixed_mlp_moe:
```
let's remove code path if this is always used or never used

ArthurZucker commented on src/transformers/models/hunyuan_v1_moe/modular_hunyuan_v1_moe.py:
```diff
@@ -0,0 +1,358 @@
+        if config.use_mixed_mlp_moe:
+            self.shared_mlp = HunYuanMoEV1MLP(config, layer_idx=layer_idx, is_shared_mlp=True)
+        self.gate = HunYuanTopKGate(config, layer_idx=layer_idx)
+        self.experts = nn.ModuleList(
+            [HunYuanMoEV1MLP(config, layer_idx=layer_idx, is_shared_mlp=False) for _ in range(self.num_experts)]
```
modulelist is worst than having 3d weights will help rather than modulist ! Don't you want to fix this to look more like gpt oss / llama4 ? will help with vllm as well

ArthurZucker commented on src/transformers/models/hunyuan_v1_moe/modular_hunyuan_v1_moe.py:
```diff
@@ -0,0 +1,358 @@
+
+        reshaped_input = hidden_states.reshape(-1, hidden_size)
+
+        dispatched_input = torch.einsum("sec,sm->ecm", dispatch_mask.type_as(hidden_states), reshaped_input)
```
no einsums please, performances are not guaranteed,

ArthurZucker reviewed (commented):
Overall LGTM, but the MoE part should be standardized as much as possible!

yjc9696 committed:
Fix moe (#7)

* fix some moe code

* fix einsum

yjc9696 commented:
> Overall LGTM, but the MoE part should be standardized as much as possible!

Hello, I've completed most of the required modifications. However, implementing the expert layer poses significant challenges for me, as it involves complex adjustments to the shape and distribution of model weights. I've referenced the relevant code from LLaMA-4, and we may consider adopting this approach in the next model version upgrade. Thank you. Could you please review if there are any other issues?

yjc9696 committed:
try top1

yjc9696 committed:
use top1

yjc9696 committed:
Fix rotary (#8)

* fix rotary

* fix modeling

yjc9696 committed:
Merge branch 'main' into hunyuan_opensource

yjc9696 committed:
fix modular

ArthurZucker commented on src/transformers/models/hunyuan_v1_moe/modular_hunyuan_v1_moe.py:
```diff
@@ -0,0 +1,365 @@
+
+
+def topkgating(logits: Tensor, topk: int):
+    if topk == 1:
```
isn't top_k always 1?

ArthurZucker commented on src/transformers/models/hunyuan_v1_moe/modular_hunyuan_v1_moe.py:
```diff
@@ -0,0 +1,365 @@
+        """Implements Top1Gating on logits."""
+        # everything is in fp32 in this function
+        logits = logits.float()
+        gates = F.softmax(logits, dim=1)
```
any reason not to use gpt_oss or mixtral gating? should be absolutely equivalent + its now fairly standard!

ArthurZucker replied:
We also have other implementations with token capacity as well!

ArthurZucker commented on src/transformers/models/hunyuan_v1_moe/modular_hunyuan_v1_moe.py:
```diff
@@ -0,0 +1,365 @@
+        reshaped_input_expanded = reshaped_input.unsqueeze(1).unsqueeze(1)  # (s, 1, 1, m)
+        dispatched_input = (dispatch_mask_expanded * reshaped_input_expanded).sum(dim=(0))  #  (s, m)
+
+        chunks = dispatched_input.chunk(self.num_experts, dim=0)
+        expert_outputs = []
+        for chunk, expert in zip(chunks, self.experts):
+            expert_outputs.append(expert(chunk))
+
+        expert_output = torch.cat(expert_outputs, dim=0)
+        # combined_output = torch.einsum("sec,ecm->sm", combine_weights.type_as(hidden_states), expert_output)
+        combine_exp = combine_weights.type_as(hidden_states).unsqueeze(3)  # (s, e, c, 1)
+        expert_exp = expert_output.unsqueeze(0)  # (1, e, c, m)
+        combined_output = (combine_exp * expert_exp).sum(dim=(1, 2))  # (s, m)
```
well, this is rigurously equivalent to the approach we have in llama4 with scattering that uses the dispatch index, I don't think there is a difference with llama4 here no? let's standardize please

ArthurZucker reviewed (approved):
Last nits on the MoE, it's pretty standard now so it is important to isolate what you are adding / what is new! From what I can see it might be the token capacity ? 
If not then let's just use what we already have standardize, know passes compile / potentially TP etc!

ArthurZucker commented:
run-slow: auto, hunyuan_v1_dense, hunyuan_v1_moe

yjc9696 committed:
fix testcode

yjc9696 commented:
run-slow: auto, hunyuan_v1_dense, hunyuan_v1_moe

ArthurZucker commented on tests/models/hunyuan_v1_moe/test_modeling_hunyuan_v1_moe.py:
Do you know why you had to do this ? Happy to help fix

yjc9696 replied:
Since the tokenizer part is not intended to be included in the open-source code, the approach of using trust_remote_code is adopted.

ArthurZucker replied:
I am not sure I understand, how are the users gonna be able to use the model if they can't use the tokenizer?

yjc9696 replied:
This is a history problem. The script you provided has already been used in our dense series models, but the MoE model was open-sourced before that. As a result, the tiktoken-tokenizer file is placed in the model file directory, so user can use the trust_remote_code approach for compatibility.

yjc9696 replied:
OK,I know what you meaning, we remove this test case until our model ready for fast tokenizer

yjc9696 replied:
We have modified the tokenizer in the model card by removing the trust_remote_code dependency and re-added the corresponding tests.

yjc9696 committed:
remove A13B unit test

yjc9696 committed:
Fix moe v1 (#9)

fix moe & gate

yjc9696 committed:
Fix gate norm (#10)

* add norm_topk_prob

yjc9696 committed:
Fix testcase (#11)

* fix&skip test

yjc9696 committed:
Fix testcase (#12)


* skip testcase

ArthurZucker commented on src/transformers/models/hunyuan_v1_moe/modular_hunyuan_v1_moe.py:
```diff
@@ -228,62 +127,72 @@ def forward(self, hidden_states):
-        bsz, seq_len, hidden_size = hidden_states.shape
+        self.shared_mlp = HunYuanMoEV1MLP(config, layer_idx=layer_idx, is_shared_mlp=True)
 
+    def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
```
a lot better thanks! 🚀

ArthurZucker reviewed (approved):
Kudos! My last comment in on `norm_topk_prob` if it is always true, let's hardcode to remove codepathes!

yjc9696 committed:
Fix norm topk (#13)

* hardcode norm_topk_prob

* fix testcase

yjc9696 commented:
> Kudos! My last comment in on `norm_topk_prob` if it is always true, let's hardcode to remove codepathes!

already fix done~
